KAFKA-8523 Enabling InsertField transform to be used with tombstone events

gunnarmorling opened the pull request:
Fixes https://issues.apache.org/jira/browse/KAFKA-8523

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

gunnarmorling committed:
KAFKA-8523 Avoiding raw type usage

gunnarmorling commented:
Hey @rhauch, a minor fix to the `InsertField`SMT -- Feedback welcome!

ftardif commented:
So, does an insert config on a tombstone <k,null> will still produce a <k, null> (hence ignored), or will it produce a <k, v> where v contains only the inserted fields. If it does not propagate the null value, I am concerned it will skew the elasticsearch connector config `behavior.on.null.values` 
that can be used to explicitly delete index entries on null.

![image](https://user-images.githubusercontent.com/221462/65040519-9006f000-d922-11e9-9968-6b8b86da7a9f.png)

rhauch commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/InsertField.java:
```diff
@@ -16,6 +16,15 @@
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
```
Nit: let's avoid moving all of these imports around. Our convention is to place the `java` and static imports at the end, and moving them unnecessarily just complicates maintenance.

gunnarmorling replied:
Reverted these.

rhauch commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/InsertFieldTest.java:
```diff
@@ -104,11 +104,64 @@ public void schemalessInsertConfiguredFields() {
+    }
+
+    @Test
+    public void insertConfiguredFieldsIntoTombstoneEvent() {
```
Nit: might be a bit clearer to add "WithoutSchema":
```suggestion
    public void insertConfiguredFieldsIntoTombstoneEventWithoutSchema() {
```

gunnarmorling replied:
Done.

rhauch reviewed (commented):
Thanks, @gunnarmorling! This looks great, but I have two minor suggestions.

ftardif commented:
before merging this PR, can we discuss the impact of applying transformation on null values? I strongly believe we should rather propagate the null values untouched even when transformations are configured. 
 
see: https://issues.apache.org/jira/browse/KAFKA-8523?focusedCommentId=16931777&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16931777

rhauch reviewed (changes requested):
@ftardif, please see the discussion on https://issues.apache.org/jira/browse/KAFKA-8523. I think @gunnarmorling, you, and I all agree that this PR should be changed to pass through and not modify tombstone records.

This PR has not yet been approved, but just to be safe I've marked this PR as `Request changes`.

ftardif commented:
@gunnarmorling are you planning to provide an adjusted PR soon in order to start testing this functionality?

gunnarmorling commented:
Yes, I'll try and send something in the course of the week. Thanks for the
reminder!

>

gunnarmorling committed:
KAFKA-8523 Gracefully handling tombstone events in InsertField SMT

gunnarmorling commented:
@rhauch, @ftardif, I've updated the PR and force-pushed as per the discussion on the JIRA issue; tombstones are returned unmodified now by the `InsertField` transformation. I have also addressed the other review remarks regarding imports and method naming.

ftardif commented:
thanks @gunnarmorling . What are the next steps to get that PR merged?

gunnarmorling commented:
Good question; I suppose approval by @rhauch or any other committer should do the trick.

rhauch reviewed (commented):
LGTM. Thanks, @gunnarmorling!

I plan to merge shortly.

gunnarmorling commented:
Excellent! To be sure, would that release be 2.4? Is there a way to learn
about the timelines for code freezes for future reference? Thanks again!

>

rhauch commented:
@gunnarmorling yes, I hope to get this into the AK 2.4.0 release. For information about release plans, checkout the dev list email thread with subject `[DISCUSS] Apache Kafka 2.4.0 release` (or the specific release you're interested in), which points to the [AK 2.4.0 release plan](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=125307901).

rhauch reviewed (approved):
LGTM.

rhauch commented:
Merged back as far back as the 1.0 branch. See https://issues.apache.org/jira/browse/KAFKA-8523 for details.

gunnarmorling commented:
Excellent; thanks for merging and backporting, Randall!

>
